src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs

```diff
@@ -1105,20 +1108,27 @@ private StackValue CreateDSObject(object obj, ProtoCore.Runtime.Context context,
         {
             //We are here, because we want to create DS object of user defined type.
             var runtimeCore = dsi.runtime.RuntimeCore;
-            var classTable = runtimeCore.DSExecutable.classTable;
             Type objType = GetPublicType(obj.GetType());
-            int type = classTable.IndexOf(GetTypeName(objType));
-            //Recursively get the base class type if available.
-            while (type == -1 && objType != null)
+
+            if (mCachedObjType != objType)
             {
-                objType = objType.BaseType;
-                if (null != objType)
-                    type = classTable.IndexOf(GetTypeName(objType));
+                mCachedObjType = objType;
+
+                var classTable = runtimeCore.DSExecutable.classTable;
+                mCachedType = classTable.IndexOf(GetTypeName(objType));
+                
+                //Recursively get the base class type if available.
+                while (mCachedType == -1 && objType != null)
+                {
+                    objType = objType.BaseType;
+                    if (null != objType)
+                        mCachedType = classTable.IndexOf(GetTypeName(objType));
+                }
             }
 
             MetaData metadata;
-            metadata.type = type;
-            StackValue retval = runtimeCore.RuntimeMemory.Heap.AllocatePointer(classTable.ClassNodes[type].Size, metadata);
+            metadata.type = mCachedType;
+            StackValue retval = runtimeCore.RuntimeMemory.Heap.AllocatePointer(0, metadata);
```

Comment:
so now we allocate an empty pointer to the object? Does this pointer actually point anywhere?

Reply:
Eventually we use this pointer StackVakue in the CLRObjectMap and DSObjectMap to retrieve or dispose the return object